virus-total-provider/filehandler.go

```diff
@@ -0,0 +1,225 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"time"
+)
+
+type fileHandler struct {
+	apiKey           string
+	baseURL          string
+	failOnFailures   bool
+	failOnSuspicious bool
+	failOnTimeout    bool
+}
+
+type fileScanRequest struct {
+	Contents []byte `json:"contents"`
+}
+
+func (f *fileHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var req fileScanRequest
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		http.Error(w, fmt.Sprintf("failed to decode request body: %v", err), http.StatusBadRequest)
+		return
+	}
+
+	uploadURL, err := f.getUploadURL(r.Context())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	analysisID, err := f.uploadFile(r.Context(), uploadURL, req.Contents)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	passed, err := f.pollForResult(r.Context(), analysisID)
+	if err != nil {
+		if m := (*maliciousErr)(nil); errors.As(err, &m) {
+			http.Error(w, m.Error(), http.StatusForbidden)
+			return
+		}
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if !passed {
+		http.Error(w, "file was flagged as malicious", http.StatusForbidden)
+	}
+}
+
+func (f *fileHandler) getUploadURL(ctx context.Context) (string, error) {
+	var resp urlResponse
+	return resp.Data, doMultipartRequest(ctx, f.apiKey, f.baseURL+"/files/upload_url", http.MethodGet, nil, &resp)
+}
+
+func (f *fileHandler) uploadFile(ctx context.Context, url string, contents []byte) (string, error) {
+	var resp fileUploadResponse
+	return resp.Data.ID, doMultipartRequest(ctx, f.apiKey, url, http.MethodPost, map[string]io.Reader{
+		"file": bytes.NewReader(contents),
+	}, &resp)
+}
+
+func (f *fileHandler) pollForResult(ctx context.Context, analysisID string) (bool, error) {
+	var (
+		resp analysisResult
+		err  error
+	)
+	ticker := time.NewTicker(time.Second)
```

Comment:
want to make sure, polling every second will be not aggressive in their api?

Reply:
Yeah, I don't know. Should we do something different do you think?

Reply:
Oh, I was thinking maybe like every 5/10 second to make sure we don't get rate limited. But it is not a problem then that's probably fine.